loader/src/sources/riteaid/api.js

```diff
@@ -185,11 +301,26 @@ async function checkAvailability(handler, options) {
 
   let results = [];
   for (const state of states) {
-    let stores;
+    // Sentry's withScope() doesn't work for async code, so we have to manually
+    // track the context data we want to add. :(
+    const errorContext = { state, source: "Rite Aid API" };
+
+    const stores = [];
     try {
-      stores = await queryState(state, rateLimit);
+      const rawData = await queryState(state, rateLimit);
+      for (const rawLocation of rawData) {
+        Sentry.withScope((scope) => {
+          scope.setContext("context", errorContext);
+          scope.setContext("location", { id: rawLocation.id });
+          try {
+            stores.push(formatStore(rawLocation));
+          } catch (error) {
+            warn(error);
+          }
+        });
+      }
     } catch (error) {
-      warn(error, { state, source: "Rite Aid API" }, true);
+      warn(error, errorContext, true);
```

Comment:
This is kind of ugly, but `withScope()` can’t handle async code. Definitely open to changes here if you have good thoughts on better patterns, or nice sugar.

Sentry has built-in support for managing async stuff with domains (for example, [the `express` plugin uses them](https://github.com/getsentry/sentry-javascript/blob/f89a842c007e894b1c9f25db604e00d7753bd4e2/packages/node/src/handlers.ts#L412-L447)), *but domains are a deprecated feature* that are going away at some point in Node.js (I think in favor of Async Hooks). So I’m hesitant to build a fancy abstraction on them.

There are also some Sentry issues tracking better async support:
- https://github.com/getsentry/sentry-javascript/issues/3660
- https://github.com/getsentry/sentry-javascript/issues/3751
- https://github.com/getsentry/sentry-javascript/issues/4071

Reply:
Yeah, I'm not sure it's worthwhile trying to build a generalized solve here, or at least not yet. The code as written reads fine to me and doesn't seem overly messy.